src/tests/Microsoft.PowerFx.Interpreter.Tests/RecalcEngineTests.cs

```diff
@@ -274,6 +278,95 @@ public static NumberValue Execute(NumberValue x, NumberValue y)
+            foreach (var span in spans)
+            {
+                referencedIdentifiers.Add(expressionText.Substring(span.Min, span.Lim - span.Min));
+            }
+            return referencedIdentifiers;
+        }
+
+        [Fact]
+        public void UnknownVariables()
+        {
+            var engine = new RecalcEngine();
+            var expr = "3*2+x+y";
+            var result = engine.Check(expr,
+                new RecordType().Add(
+                    new NamedFormulaType("x", FormulaType.Number)));
+
+            Assert.False(result.IsSuccess);
+            var referencesIdentifierStrings = ReferencedIdentifersSet(result.ReferencedIdentifiers, expr);
+            Assert.Equal(2, result.ReferencedIdentifiers.Count);
+            Assert.Contains("x", referencesIdentifierStrings);
+            Assert.Contains("y", referencesIdentifierStrings);
+        }
+
+        [Fact]
+        public void UnknownDottedVariable()
+        {
+            var engine = new RecalcEngine();
+            var expr = "3*2+user.age";
+            var result = engine.Check(expr,
+                new RecordType().Add(
+                    new NamedFormulaType("user", new RecordType())));
+
+            Assert.False(result.IsSuccess);
+            Assert.Single(result.ReferencedIdentifiers);
+            var referencesIdentifierStrings = ReferencedIdentifersSet(result.ReferencedIdentifiers, expr);
+            Assert.Equal("user.age", referencesIdentifierStrings.First());
+        }
+
+        [Fact]
+        public void UnknownDottedDottedVariable()
+        {
+            var engine = new RecalcEngine();
+            var expr = "3*2+user.car.age";
+            var result = engine.Check(expr,
+                new RecordType().Add(
+                    new NamedFormulaType("user", new RecordType())));
+
+            Assert.False(result.IsSuccess);
+            Assert.Single(result.ReferencedIdentifiers);
+            var referencesIdentifierStrings = ReferencedIdentifersSet(result.ReferencedIdentifiers, expr);
+            Assert.Equal("user.car.age", referencesIdentifierStrings.First());
+        }
+
+        [Fact]
+        public void FunctionVariable()
+        {
+            var engine = new RecalcEngine();
+            var expr = "sum(user.age)";
+            var result = engine.Check(expr,
+                new RecordType().Add(
+                    new NamedFormulaType("user", new RecordType())));
+
+            Assert.False(result.IsSuccess);
+            Assert.Single(result.ReferencedIdentifiers);
+            var referencesIdentifierStrings = ReferencedIdentifersSet(result.ReferencedIdentifiers, expr);
+            Assert.Equal("user.age", referencesIdentifierStrings.First());
+        }
+
+        [Fact]
+        public void ManyVariables()
+        {
+            var engine = new RecalcEngine();
+            var expr = "(27/x * test(user.age) - global.shoeSize) ^ exponent";
+            var result = engine.Check(expr,
```

Comment:
What about something a bit more complex?

`With(user, age+1)`, where this is using the field "age" of the user record. That'll return {"user", "age"} with the current implementation, is that the behavior we want?

Reply:
We're currently requiring fully qualified names, so you'd have to say "user.age" to get the age field of the user record. So I think this matches the intended behavior.

Reply:
That's not exactly how With and record scopes in general. User is a record, so it's fields are usable in lambdas of functions like With, ForAll, Filter, ...

Reply:
Or, let's say you have an object with two user children:

`If(PickFirst, Foo.User1, Foo.User2).Name`